docs/madr/decisions/031-automatic-rechable-services.md

```diff
@@ -0,0 +1,263 @@
+# Automatic reachable services
+
+* Status: accepted
+
+Technical Story: https://github.com/kumahq/kuma/issues/6551
+
+## Context and Problem Statement
+
+With large enough Mesh (more than 1000 services) processing the whole cluster state for each proxy is not possible due to performance reasons (what is the boundary that we're hitting, is it the CPU of the CP / DP, can we have flame graphs showing hot spots?).
+This is why a feature called [reachable services](https://github.com/kumahq/kuma-website/blob/045ff9868acdb4b322759dde3f8053f41948100f/app/_src/production/upgrades-tuning/fine-tuning.md#L8-L7) was introduced.
+Reachable services is a step in the right direction but defining the whole service graph manually by the service owners or the platform team is a big ask.
+
+How do we ensure a smooth Mesh migration (keeping in mind the performance and minimising manual labour) for large systems?
+
+## Glossary
+
+- MTP - [MeshTrafficPermission](https://kuma.io/docs/2.4.x/policies/meshtrafficpermission/#meshtrafficpermission-(beta))
+
+## Decision Drivers 
+
+* feasibility of implementation for 2.5 release
+* minimize the manual labour needed
+* smooth migration to Mesh for large systems without downtime or service communication issues
+
+## Considered Options
+
+* Using MeshTrafficPermission to generate reverse graph that would feed into reachable services
+* Dynamically load clusters using ODCDS
+* Using MeshHTTPRoute / MeshTCPRoute to populate the clusters
+
+## Decision Outcome
+
+Chosen option: "Using MeshTrafficPermission to generate reverse graph that would feed into reachable services"
+
+### Positive Consequences
+
+* not invasive, plug and play, can be removed if proven not efficient enough
+* possible to implement in the given time frame
+
```

Comment:
Another positive consequence is that it still works with reachable services and folks can mix and match both options.

Reply:
added